### PR TITLE
Search feature

### PR DESCRIPTION
**Pull request by CulturalProfessor**
Fixes(#32 )

### CODE REVIEW

File: imports/models/BoardCollection.ts [hunk 1]
```diff
@@ -10,6 +10,9 @@ export type BoardRecord = IBoard & {
 }
 
 export class BoardCollectionClass extends BaseCollection<BoardRecord> {
+  searchBoards() {
```
Comment (shiqimei) [inline imports/models/BoardCollection.ts:13]: make it a public method, pass `keyword: string` as parameter. use

```typescript
db.collection.find({
  $or: [
    {
      "elements.text": {
        $regex: "hello",
        $options: "i"
      }
    },
    {
      "title": {
        $regex: "hello",
        $options: "i"
      }
    }
  ]
})
```

to filter boards which `title` or text elements contains search keywords.

Comment (shiqimei) [reply imports/models/BoardCollection.ts:13]: check out this example: https://mongoplayground.net/p/D4PPy8eg38A

to see how i query boards which text elements or title contains "hello" keyword

hope this helps

File: imports/services/SearchService.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    const boards = BoardCollection.raw.find({ userId }).fetch()
+
+    const options = {
+      keys: ['name', 'description'],
+      threshold: 0.2 // The minimum score required for a match to be considered a match
+    }
+
+    const fuse = new Fuse(boards, options)
```
Comment (shiqimei) [inline imports/services/SearchService.ts:34]: fuse.js is great for fuzzy search in memory array. for mongodb, use `$regex` for exact search is enough currently. refer to my comments above.

remove this dependency please, if we have more time imrpove search feature, let's try to reintroduce fuse.js later.

File: imports/subscriptions/SearchSubscriber.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    globalEventEmitter.on(SearchEvents.SearchChanged, onSearchChanged)
+    SearchSubscriber.subscribe()
+    return () => {
+      globalEventEmitter.off(SearchEvents.SearchChanged, onSearchChanged)
+    }
+  }, [])
+  return boards
+}
+
+class SearchSubscriberClass {
```
Comment (shiqimei) [inline imports/subscriptions/SearchSubscriber.ts:27]: maybe for search results, we can use `searchBoards` in BoardService. we don't need to introduce `SearchSubscriber` to make the search results "reactive". it's too heavy.

we can call `Services.get('boards').searchBoards` when use typing search keywords in input box. I think this way is more simple.

**Review by shiqimei — CHANGES_REQUESTED** [review]

File: imports/services/ServiceManagerServer.ts [hunk 4]
```diff
@@ -21,6 +22,7 @@ class ServiceManagerServerClass {
     this._services['board_favorite'] = new BoardFavoriteService()
     this._services['files'] = new FilesService()
     this._services['redirect'] = new RedirectService()
+    this._services['search'] = new SearchService()
```
Comment (shiqimei) [inline imports/services/ServiceManagerServer.ts:25]: we can add `searchBoards` in board service, no need to add an extra service.

File: imports/pages/Board/FileExplorerPanel/FindAnything/FindAnything.module.scss [hunk 5]
```diff
@@ -1,27 +1,67 @@
+
 .FindAnything {
-	background: #e7eaf1;
```
[the remark this answers is not in the document]
Comment (CulturalProfessor) [reply imports/pages/Board/FileExplorerPanel/FindAnything/FindAnything.module.scss:2]: yes

**Comment by shiqimei** [comment]
Closes #32

**Comment by shiqimei** [comment]
LGTM! Thanks for your work. Let's continue on #42 .

**Comment by CulturalProfessor** [comment]
> LGTM! Thanks for your work. Let's continue on #42 .

Yeah let's improve it !!

**Pull request closed without merging**